src/core/ext/transport/binder/transport/binder_transport.cc

```diff
@@ -157,21 +157,25 @@ static void perform_stream_op(grpc_transport* gt, grpc_stream* gs,
   }
   if (op->send_message && gbs->cancellation_error.ok()) {
     gpr_log(GPR_INFO, "send_message");
-    grpc_slice s;
-    bool next_result =
-        op->payload->send_message.send_message->Next(SIZE_MAX, nullptr);
-    gpr_log(GPR_INFO, "next_result = %d", static_cast<int>(next_result));
-    op->payload->send_message.send_message->Pull(&s);
-    auto* p = GRPC_SLICE_START_PTR(s);
-    int len = GRPC_SLICE_LENGTH(s);
-    std::string message_data(reinterpret_cast<char*>(p), len);
+    size_t remaining = op->payload->send_message.send_message->length();
+    std::string message_data;
+    do {
+      grpc_slice message_slice;
+      GPR_ASSERT(
+          op->payload->send_message.send_message->Next(SIZE_MAX, nullptr));
+      op->payload->send_message.send_message->Pull(&message_slice);
```

Comment:
Is it guaranteed that the Pull will success? Do we need error handling here?

Reply:
Actually, I should ask why we can confidently assert Next() will return true

Reply:
And what happen when `remaining` is initially 0

Reply:
`Pull()` might fail and we should probably just cancel the stream in this case.

`Next()` should returns true because we know that there's still messages to come (judging from `remaining`). `inproc_transport.cc` also asserts `Next()` to return true. I think `remaining` should also be positive in the beginning because why sending an empty message?

Reply:
I've added an assertion that `error == GRPC_ERROR_NONE`. We can call `cancel_stream_locked()` after merging https://github.com/grpc/grpc/pull/27182.

What do you think?

Reply:
> Next() should returns true because we know that there's still messages to come

https://github.com/grpc/grpc/blob/ca945a58e9ccff69df9815bdb657c4b32e2e6b37/src/core/lib/transport/byte_stream.h#L45-L49 Says "false if the bytes will be available asynchronously" so I'm not sure if that case can happen or not

> I think remaining should also be positive in the beginning because why sending an empty message?

IIRC there is a special test case that tests if the transport supports sending empty message? Change your do-while loop to while loop should be enough to account for that situation

Reply:
Ok to add a comment above the assert and merge this first, so that reader can know the situation easily if the assertion fails in the future

Reply:
Sure. Done.